### DISCUSSION TITLE
how to serve different schemas based on user

### DISCUSSION THREAD

**Discussion by alexrohleder** [apollographql/apollo-server#5684] [Q&A] [answered]
hey,

I have a relational database that I want to expose via GraphQL.

An item will have attributes, everything is scoped for a user. The tables looks like `items -> items_to_attributes -> attributes`.

I wonder if I can *dynamically generate the schema based on creation, update or deletion of attributes, for each user.*

For example:

- user1 has attr1, attr2
- user2 has attr3, attr4, attr5

I want them to be able to query like this:

- user1 `
query {
  item {
    attr1,
    attr2
  }
}
`
- user2 `
query {
  item {
    attr3,
    attr4,
    attr5
  }
}
`

thanks in advance :)

**Comment by glasser** [comment, marked as answer]
We don't support dynamic schemas. There's a lot of complexity there around how plugins, reporting, etc would deal with a dynamic schema. See eg https://github.com/apollographql/apollo-server/pull/5071#issuecomment-902059802

For what it's worth, having a super dynamic schema will make it challenging to do things like code generation, which is one of the main strengths of GraphQL.

**Reply by alexrohleder** [reply]
thanks for the explanation and the code generation example :) @glasser

I ended up using aliases to name several attributes:

```
query {
  item {
    attr1: attr(id: 1)
    attr2: attr(id: 2)
  }
}
```
